File: C++/Recursion_Backtracking_Medium_Permutations_46.cpp
// Solution using a visited_array:

class Solution {
public:

    void helper(vector<int>& nums, vector<vector<int>>& result, vector<int>& ds, vector<int>& visited_array)
    {
        if(ds.size() == nums.size())
        {
            result.push_back(ds);
            return;
        }

        for(int i = 0; i < nums.size(); i++)
        {
            if(!visited_array[i])
            {
                ds.push_back(nums[i]);
                visited_array[i]++;
                helper(nums, result, ds, visited_array);
                ds.pop_back();
                visited_array[i] = 0;
            }
        }
    }

    vector<vector<int>> permute(vector<int>& nums)
    {   
        vector<vector<int>> result;
        vector<int> ds;
        vector<int> visited_array(nums.size(), 0);

        helper(nums, result, ds, visited_array);

        return result;
        
    }
};

// Solution using swapping:

class Solution {
public:

    void helper(vector<int>& nums, int index, vector<vector<int>>& answer)
    {
        if(index == nums.size())
        {
            answer.push_back(nums); 
            return;
        }

        for(int i = index; i < nums.size(); i++)
        {
            swap(nums[index], nums[i]);
            helper(nums, index + 1, answer);
            swap(nums[index], nums[i]);
        }
    }

    vector<vector<int>> permute(vector<int>& nums) 
    {
        int index = 0;

        vector<vector<int>> answer;

        helper(nums, index, answer);

        return answer;
    }
};

File: Python/Recursion_Backtracking_Medium_Permutations_46.py
# Solution using a visited_array:

class Solution:

    def helper(self, nums, result, ds, visited_array):
        if len(ds) == len(nums):
            result.append(ds[:])
            return
        
        for i in range(0, len(nums)):
            if not visited_array[i]:
                ds.append(nums[i])
                visited_array[i] += 1
                self.helper(nums, result, ds, visited_array)
                ds.pop()
                visited_array[i] = 0


    def permute(self, nums: List[int]) -> List[List[int]]:
        result = []
        ds = []
        visited_array = [0] * len(nums)

        self.helper(nums, result, ds, visited_array)

        return result
        
# Solution using swapping: 

class Solution:

    def helper(self, nums, index, answer):
        if index == len(nums):
            answer.append(nums[:])
            return
        
        for i in range(index, len(nums)):
            nums[index], nums[i] = nums[i], nums[index]
            self.helper(nums, index + 1, answer)
            nums[index], nums[i] = nums[i], nums[index]

    def permute(self, nums: List[int]) -> List[List[int]]:
        answer = []
        index = 0

        self.helper(nums, index, answer)

        return answer
        

